### PR TITLE
added remote auth module

### PR DESCRIPTION
**Pull request by JulianWielga**
separated authentication strategies
better dynamic modules handling
refactoring of NussknackerInitializer

### CODE REVIEW

File: ui/client/containers/Auth/strategies/FallbackStrategy.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import SystemUtils from "../../../common/SystemUtils"
+import {Strategy, StrategyConstructor} from "../Strategy"
+
+export const FallbackStrategy: StrategyConstructor = class FallbackStrategy implements Strategy {
```
Comment (mproch) [inline ui/client/containers/Auth/strategies/FallbackStrategy.ts:4]: Let's add comment that it usually amounts to BASIC?

File: ui/client/containers/Auth/Strategy.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  new(
+    settings: AuthenticationSettings,
+    onError?: (error: AuthErrorCodes) => void,
+  ): Strategy,
+}
+
+export interface Strategy {
+  Wrapper?: (props: PropsWithChildren<unknown>) => JSX.Element,
+
+  inteceptor?<E>(error: E): Promise<unknown>,
```
Comment (mproch) [inline ui/client/containers/Auth/Strategy.ts:15]: can we add comment when/why is it needed?

File: ui/client/containers/Auth/strategies/RemoteAuthStrategy/externalAuthModule.d.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
+import {ComponentType, PropsWithChildren} from "react"
```
Comment (mproch) [inline ui/client/containers/Auth/strategies/RemoteAuthStrategy/externalAuthModule.d.ts:1]: I think we need some comments regarding:
- that this interface can be implemented by external modules
- short description how workflow for externalAuthModule works, what is possible etc.

File: ui/client/containers/ExternalLib/ExternalModule.tsx [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import {lazy} from "@loadable/component"
+import React, {PropsWithChildren, useMemo} from "react"
+import {LibContextProvider} from "./store"
+import {createScript, loadComponent, splitUrl} from "./tools"
+import {Module, ModuleUrl} from "./types"
+
+export function ExternalModule<M extends Module>(props: PropsWithChildren<{url: ModuleUrl}>): JSX.Element {
```
Comment (mproch) [inline ui/client/containers/ExternalLib/ExternalModule.tsx:7]: Do I understand correctly, that this is central point of accessing module federation? Can we add some comment to this class? E.g. this is the place to look for if you want to use externalmodules in NK, see remoteTab for example

**Review by mproch — COMMENTED** [review]
Looks great, I just miss some comments here and there...

File: ui/client/containers/ExternalLib/tools/splitUrl.tsx [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+export function splitUrl(url: ModuleUrl): [ModuleUrl, ModuleString, ScriptUrl, ScopeString, PathString] {
+  const [module, script] = url.split("@")
+  const [scope] = module.split("/")
+  const path = module.replace(scope, ".")
+
+  if (!scope || !script.match(/^http.*\.js$/)) {
+    throw new Error("invalid remote module url")
+  }
+
+  console.table([url, module, script, scope, path])
```
[the remark this answers is not in the document]
Comment (JulianWielga) [reply ui/client/containers/ExternalLib/tools/splitUrl.tsx:12]: no ;)

**Review by mproch — APPROVED** [review]

File: ui/client/containers/errors/InitializeError.tsx [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  description?: string,
+  buttonLabel?: string,
+  showButton?: boolean,
+  buttonOnClick?: () => void,
+}
+
+function InitializeError(props: ErrorProps): JSX.Element {
+  const defaultOnClick = useCallback(() => window.location.reload(), [])
+  const {
+    message = "Application Unexpected Error",
```
Comment (lciolecki) [inline ui/client/containers/errors/InitializeError.tsx:14]: Missing translations.

File: ui/client/reducers/settings.ts [hunk 7]
```diff
@@ -12,8 +18,8 @@ export type SettingsState = {
 }
 
 export type AuthenticationSettings = {
-  backend: string,
-  authorizeUrl?: URL,
+  backend: AuthBackends,
+  authorizeUrl?: string,
```
Comment (lciolecki) [inline ui/client/reducers/settings.ts:22]: Why not URL?

File: ui/client/containers/Auth/Strategy.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import {AuthErrorCodes} from "./AuthErrorCodes"
+
+export interface StrategyConstructor {
+  new(
+    settings: AuthenticationSettings,
+    onError?: (error: AuthErrorCodes) => void,
+  ): Strategy,
+}
+
+export interface Strategy {
```
[the remark this answers is not in the document]
Comment (JulianWielga) [reply ui/client/containers/Auth/Strategy.ts:12]: as you can see it's packed in directory ("ui/client/containers/Auth/Strategy.ts"), and used only by modules within this directory. this way we'll end with 1000 characters names

File: ui/client/reducers/settings.ts (hunk 7 above)
Comment (JulianWielga) [reply ui/client/reducers/settings.ts:22]: to avoid changes outside frontend part. when settings/props will change this will also change.

**Pull request merged**